tracer/src/Datadog.Trace/Iast/Propagation/PropagationModuleImpl.cs

```diff
@@ -0,0 +1,139 @@
+            var taintedSelf = taintedObjects.Get(input);
+
+            if (taintedSelf == null)
+            {
+                return result;
+            }
+
+            if (offset != 0)
+            {
+                var newRanges = new Range[taintedSelf.Ranges.Length];
+                Ranges.CopyShift(taintedSelf.Ranges, newRanges, 0, offset);
+                taintedObjects.Taint(result, newRanges);
+            }
+            else
+            {
+                taintedObjects.Taint(result, taintedSelf.Ranges);
+            }
+        }
+        catch (Exception err)
+        {
+            Log.Error(err, "StringModuleImpl.PropagateTaint exception");
+        }
+
+        return result;
+    }
+
+    public static TaintedObject? GetTainted(TaintedObjects taintedObjects, object? value)
+    {
+        return value == null ? null : taintedObjects.Get(value);
+    }
+
+    /// <summary> Taints a string.substring operation </summary>
+    /// <param name="self"> original string </param>
+    /// <param name="beginIndex"> start index </param>
+    /// <param name="result"> the substring result </param>
+    /// <param name="resultLength"> Result's length </param>
+    public static void OnStringSubSequence(object self, int beginIndex, object result, int resultLength)
+    {
+        var iastContext = IastModule.GetIastContext();
+        if (iastContext == null)
+        {
+            return;
+        }
+
+        var taintedObjects = iastContext.GetTaintedObjects();
+        var selfTainted = taintedObjects.Get(self);
+        if (selfTainted == null)
+        {
+            return;
+        }
+
+        var rangesSelf = selfTainted.Ranges;
+        if (rangesSelf.Length == 0)
+        {
+            return;
+        }
+
+        var newRanges = Ranges.ForSubstring(beginIndex, resultLength, rangesSelf);
+        if (newRanges != null && newRanges.Length > 0)
+        {
+            taintedObjects.Taint(result, newRanges);
+        }
+    }
+
+    // It could potentillay happen that a StringBuilder would have incorrect ranges. Reasons for that include: a not covered stringBuilder method
+    // is used or a stringbuilder method is called through reflection. This topic has been discussed and it is an issue that could occur
+    // in different languages.
+    // This situation could affect the string builder class, but not the string class because the string methods return new instances
+    // After discussion, we assume that we can have incorrect ranges in special situations, but we should make sure that the ranges do not
+    // exceed the string length
+    public static void FixRangesIfNeeded(string result)
+    {
+        var tainted = IastModule.GetIastContext()?.GetTaintedObjects()?.Get(result);
+        var ranges = tainted?.Ranges;
+
+        if (ranges == null)
+        {
+            return;
+        }
+
+        var incorrectRanges = false;
+        List<Range> newRanges = new();
```

Comment:
nit: could lazily allocate this, only if there are incorrect ranges. Also, we know the max size, so can use `new(ranges.Length)`

Reply:
Good point! I have updated the code.